# <center>Практическое задание к уроку 5

__1.__ Найти с помощью NumPy SVD для матрицы

$\begin{pmatrix}
1 & 2 & 0\\ 
0 & 0 & 5\\ 
3 & -4 & 2\\ 
1 & 6 & 5\\ 
0 & 1 & 0
\end{pmatrix}$.


Используем методы модуля linalg библиотеки  numpy:

In [16]:
import numpy as np
np.set_printoptions(precision=2, suppress=True) #устанавливаем параметры вывода чисел с плавающей запятой.

In [10]:
A=np.array([[1,2,0],[0,0,5],[3,-4,2],[1,6,5],[0,1,0]])

In [11]:
U, s, W = np.linalg.svd(A)

In [12]:
D = np.zeros_like(A, dtype=float)
D[np.diag_indices(min(A.shape))] = s

In [13]:
print(f'Матрица A:\n{A}')
print(f'Матрица U:\n{U}')
print(f'Матрица D:\n{D}')
print(f'Матрица V:\n{W.T}')

Матрица A:
[[ 1  2  0]
 [ 0  0  5]
 [ 3 -4  2]
 [ 1  6  5]
 [ 0  1  0]]
Матрица U:
[[ 0.17  0.16 -0.53 -0.8  -0.16]
 [ 0.39 -0.53  0.61 -0.43  0.03]
 [-0.14 -0.82 -0.52  0.14  0.07]
 [ 0.89  0.06 -0.25  0.38 -0.06]
 [ 0.08  0.11 -0.08 -0.11  0.98]]
Матрица D:
[[8.82 0.   0.  ]
 [0.   6.14 0.  ]
 [0.   0.   2.53]
 [0.   0.   0.  ]
 [0.   0.   0.  ]]
Матрица V:
[[ 0.07 -0.37 -0.93]
 [ 0.72  0.67 -0.21]
 [ 0.69 -0.65  0.31]]


In [14]:
U@D@W

array([[ 1.,  2.,  0.],
       [ 0., -0.,  5.],
       [ 3., -4.,  2.],
       [ 1.,  6.,  5.],
       [-0.,  1.,  0.]])

__2.__ Для матрицы из предыдущего задания найти:

   а) евклидову норму;
    
   б) норму Фробениуса.

---
a) Евклидова норма матрицы равна евклидовой норме диагональной матрицы из ее сингулярных чисел $D$. Максимальное значение полученного отношения будет равно максимальному сингулярному числу $\mu_{max}$, и, принимая во внимание факт сортировки по убыванию сингулярных чисел, получим

$\left \| A \right \|_{E}=\mu_{1}\\
\Rightarrow \left \| A \right \|_{E}= 8.82$.  

Проверим:

In [17]:
np.linalg.norm(A, ord=2)

8.824868854820444

---
б)  Норма Фробениуса определяется по формуле:  

$\left \| A \right \|_{F}=\sqrt{\sum_{i=1}^{m}\sum_{j=1}^{n}a_{ij}^{2}}$.  

Напишем функцию для нахождения:

In [19]:
def my_frobenius(matrix):
    sum_square=0
    for i in range(len(matrix)):
        for j in range(len(matrix[0])):
            sum_square+=matrix[i][j]**2
    return np.sqrt(sum_square)

print(f'Норма фробениуса для матрицы A:\n{my_frobenius(A)}')

Норма фробениуса для матрицы A:
11.045361017187261


Кроме того, когда известно сингулярное разложение матрицы, ее норма Фробениуса вычисляется как

$\left \| A \right \|_{F}=\sqrt{\sum_{k=1}^{r}\mu_{k}^{2}}$  

найдем:

In [21]:
U,s,W=np.linalg.svd(A)
s

array([8.82, 6.14, 2.53])

In [24]:
frob_norm=np.sqrt(np.sum(s**2))
print(f'Норма фробениуса для матрицы A:\n{frob_norm}')

Норма фробениуса для матрицы A:
11.045361017187261


Найдем стандартным методом модуля linalg:  

In [28]:
print(f"Норма фробениуса для матрицы A:\n{np.linalg.norm(A, ord='fro')}")

Норма фробениуса для матрицы A:
11.045361017187261
